src/stage.c

```diff
@@ -444,8 +444,13 @@ stage_request(struct view *view, enum request request, struct line *line)
 				string_ncopy(view->env->file, file, strlen(file));
 		}
 
-		view->env->ref[0] = 0;
-		view->env->goto_lineno = diff_get_lineno(view, line);
+		if (!is_initial_commit() && line->type == LINE_DIFF_DEL) {
```

Comment:
The reason for `!is_initial_commit()` here is not obvious to me, can you explain ?

Reply:
Oh well, in this case the blame will fail anyway with "Failed to load blame data". I removed it now, this is clearer.

Reply:
Ok I see what you wanted to do. What we'd need here is a way to detect an "AM" status...

Reply:
I quite agree that diff_get_old_lineno should be merged into diff_get_lineno. Something like `unsigned int diff_get_lineno(struct view *view, struct line *line, bool old)`.  Then, with stage.c modified this way, it should do it.

```patch
@@ -445,7 +445,16 @@ stage_request(struct view *view, enum request request, struct line *line)
 		}
 
 		view->env->ref[0] = 0;
-		view->env->goto_lineno = diff_get_lineno(view, line);
+		if (line->type == LINE_DIFF_DEL) {
+			const char *ls_tree_argv[] = {
+				"git", "ls-tree", "-z", "HEAD", view->env->file, NULL
+			};
+			char buf[SIZEOF_STR] = "";
+			io_run_buf(ls_tree_argv, buf, sizeof(buf), NULL, false);
+			if (buf[0])
+				string_copy(view->env->ref, "HEAD");
+		}
+		view->env->goto_lineno = diff_get_lineno(view, line, !!view->env->ref[0]);
 		if (view->env->goto_lineno > 0)
 			view->env->goto_lineno--;
 		return request;
```

Reply:
Yeah, that works nicely! Now the behavior for "AM" files is as before: the working-tree version of the file is opened in the blame-view. Not very useful/common but still better than just showing an error.

I've also added proper handling of "MM" files now - when blaming an unstaged deletion, I parse `git diff-index --cached -p HEAD -- filename` to compute the correct line number in `HEAD`.

What is still missing are uncommitted renames - blame currently has the old behavior here.
Perhaps the way to fix this is by running `git diff-index --cached -C -p HEAD` and then looking for a diff that renames the current file, though it would be nicer to reuse the status information here.